src/typeGuards.ts

```diff
@@ -0,0 +1,1368 @@
+  const typedObj = obj as FaceTuple
+  return (
+    (typedObj !== null &&
+      typeof typedObj === "object" ||
+      typeof typedObj === "function") &&
+    typeof typedObj["id"] === "string" &&
+    typeof typedObj["face"] === "object"
+  )
+}
+
+export function isHistoryStep(obj: unknown): obj is HistoryStep {
+  const typedObj = obj as HistoryStep
+  return (
+    (typedObj !== null &&
+      typeof typedObj === "object" ||
+      typeof typedObj === "function") &&
+    typeof typedObj["name"] === "string" &&
+    typeof typedObj["suppressed"] === "boolean" &&
+    typeof typedObj["unique_id"] === "string" &&
+    ((typedObj["data"] !== null &&
+      typeof typedObj["data"] === "object" ||
+      typeof typedObj["data"] === "function") &&
+      isHistoryStepType(typedObj["data"]["type"]) as boolean &&
+      isPoint3D(typedObj["data"]["point"]) as boolean ||
+      (typedObj["data"] !== null &&
+        typeof typedObj["data"] === "object" ||
+        typeof typedObj["data"] === "function") &&
+      isHistoryStepType(typedObj["data"]["type"]) as boolean &&
+      isPlane(typedObj["data"]["plane"]) as boolean &&
+      typeof typedObj["data"]["width"] === "number" &&
+      typeof typedObj["data"]["height"] === "number" ||
+      (typedObj["data"] !== null &&
+        typeof typedObj["data"] === "object" ||
+        typeof typedObj["data"] === "function") &&
+      isHistoryStepType(typedObj["data"]["type"]) as boolean &&
+      (typedObj["data"]["extrusion"] !== null &&
+        typeof typedObj["data"]["extrusion"] === "object" ||
+        typeof typedObj["data"]["extrusion"] === "function") &&
+      typeof typedObj["data"]["extrusion"]["sketch_id"] === "string" &&
+      Array.isArray(typedObj["data"]["extrusion"]["face_ids"]) &&
+      typedObj["data"]["extrusion"]["face_ids"].every((e: any) =>
+        typeof e === "number"
+      ) &&
+      typeof typedObj["data"]["extrusion"]["length"] === "number" &&
+      typeof typedObj["data"]["extrusion"]["offset"] === "number" &&
+      typeof typedObj["data"]["extrusion"]["direction"] === "string" &&
+      typeof typedObj["data"]["extrusion"]["mode"] === "string" ||
+      (typedObj["data"] !== null &&
+        typeof typedObj["data"] === "object" ||
+        typeof typedObj["data"] === "function") &&
+      isHistoryStepType(typedObj["data"]["type"]) as boolean &&
+      (typedObj["data"]["plane_description"] !== null &&
+        typeof typedObj["data"]["plane_description"] === "object" ||
+        typeof typedObj["data"]["plane_description"] === "function") &&
+      typeof typedObj["data"]["plane_description"]["PlaneId"] === "string" &&
+      typeof typedObj["data"]["width"] === "number" &&
+      typeof typedObj["data"]["height"] === "number" &&
+      (typedObj["data"]["sketch"] !== null &&
+        typeof typedObj["data"]["sketch"] === "object" ||
+        typeof typedObj["data"]["sketch"] === "function") &&
+      (typedObj["data"]["sketch"]["points"] !== null &&
```

Comment:
wow, did you type this by hand or is this autogenerated?

Reply:
Both! `pnpm type:guards` runs ts-auto-guard. It writes to `types.guard.ts` which is ignored in `.gitignore`. Then I copypasta into here and fix up the errors - git diff helps here because I can see the previous corrections. 

It gets silly little things wrong, especially arrays such as: `typeof typedObj["faces"] === "array"`. I just manually convert it to `Array.isArray...`. 

Most of the tests I've written so far are for checking the checkers! As we add runtime typechecks to the code we need to first add tests to confirm the type guards behave correctly. Otherwise it's a bit meaningless haha